cmake/modules/extensions.cmake

```diff
@@ -1593,20 +1593,17 @@ function(zephyr_build_string outvar)
   string(JOIN "_" ${outvar} ${str_segment_list} ${revision_string} ${BUILD_STR_BUILD})
 
   if(BUILD_STR_MERGE)
-    if(DEFINED BUILD_STR_BOARD_REVISION)
-      string(JOIN "_" variant_string ${str_segment_list} ${BUILD_STR_BUILD})
+    string(JOIN "_" variant_string ${str_segment_list} ${BUILD_STR_BUILD})
+
+    if(NOT "${variant_string}" IN_LIST ${outvar})
       list(APPEND ${outvar} "${variant_string}")
     endif()
-    list(POP_BACK str_segment_list)
-    while(NOT str_segment_list STREQUAL "")
-      if(DEFINED BUILD_STR_BOARD_REVISION)
-        string(JOIN "_" variant_string ${str_segment_list} ${revision_string} ${BUILD_STR_BUILD})
-        list(APPEND ${outvar} "${variant_string}")
-      endif()
-      string(JOIN "_" variant_string ${str_segment_list} ${BUILD_STR_BUILD})
-      list(APPEND ${outvar} "${variant_string}")
-      list(POP_BACK str_segment_list)
-    endwhile()
+
+    list(LENGTH str_segment_list str_segment_list_len)
+    if("${str_segment_list_len}" EQUAL "2")
+      # Include board name without SoC
+      list(APPEND ${outvar} "${BUILD_STR_BOARD}")
+    endif()
```

Comment:
initially I'm not sure we should change the behavior of `MERGE`.
Instead we should probably avoid using `MERGE` in this case, and instead update code here:
https://github.com/zephyrproject-rtos/zephyr/blob/62665eabf7d29bbdf3309e197d3d01d67b669b21/cmake/modules/configuration_files.cmake#L79-L89

That code can then do something like:
```
  zephyr_build_string(board_overlay_string
                      BOARD ${BOARD}
                      BOARD_IDENTIFIER ${BOARD_IDENTIFIER}
  )
  set(board_overlay_string "${board_overlay_string}.overlay")
  if(NOT EXISTS "${board_overlay_string}")
    string(REGEX REPLACE" "^[^/]*/" "" legacy_board_identifier "${BOARD_IDENTIFIER}"
​    zephyr_build_string(legacy_board_overlay_string
                        BOARD ${BOARD}
                        BOARD_IDENTIFIER ${legacy_board_identifier}
    )
    set(legacy_board_overlay_string "${legacy_board_overlay_string}.overlay")
    if(EXISTS "${legacy_board_overlay_string}.overlay")
        message(WARNING Overlay file using legacy board target file name structure (${legacy_board_overlay_string}) detected.
         Rename overlay to: ${board_overlay_string})
        set(board_overlay_string ${legacy_board_overlay_string})
    endif()
  endif()
  
  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
              NAMES "${APP_BOARD_DTS};${board_overlay_strings};app.overlay" SUFFIX ${FILE_SUFFIX})
```

and do similar for conf files.

After two releases, then warning code can be removed.
(We can already place the warning code in deprecated.cmake now, but that's outside the point of the example given)

Reply:
The changes here also apply to board files (i.e. in `boards`), I don't remember if it was raised an issue on github or discord or internally but an issue was raised about common files setting things that then conflicted with configuration for variants, one was for trustzone, another was for qemu, the changes here are intentionally applied to the base function.

Reply:
> The changes here also apply to board files (i.e. in boards)

Yes, and that would need to be handled in the lines just above:
https://github.com/zephyrproject-rtos/zephyr/blob/62665eabf7d29bbdf3309e197d3d01d67b669b21/cmake/modules/configuration_files.cmake#L76

and `zephyr_file()` should no longer apply `MERGE`, but do identical to Zephyr v3.6.0, which was:
https://github.com/zephyrproject-rtos/zephyr/blob/468eb56cf242eedba62006ee758700ee6148763f/cmake/modules/extensions.cmake#L2550-L2554

instead of current main:
https://github.com/zephyrproject-rtos/zephyr/blob/2396b01b1ea2b0b6a61591155508cf19cd9d367c/cmake/modules/extensions.cmake#L2607

Of course the change to use BOARD_QUALIFIERS after v3.6.0 should be kept.
Only MERGE to be removed.

Reply:
I don't think this should be done, taking the rpi_pico board, that is the board name, there is no other soc for this device and the twister identifier for this in hwmv2 is as it was in hwmv2: `rpi_pico`, the `w` variant had to have the name updated to `rpi_pico/rp2040/w` because that is required for variants but this would be a drastic change which then would invalidate (deprecate) **all** board names for in-tree boards (which are also used throughout tests and samples) which do not have the soc in the name. I would like @nashif to give feedback on if this is acceptable

Reply:
> but this would be a drastic change which then would invalidate (deprecate) all board names for in-tree boards (which are also used throughout tests and samples) which do not have the soc in the name. 

We need to make it clear how board overlays are being picked up.
Either always merge, as we do now, which clearly have drawbacks. 
Or only used the normalized board target name, where we don't merge at all, but allow legacy format for a given time (that is file names without the SoC qualifier).

A combination where you would just allow SoC to be omitted but require the CPU set and variant parts will create a mess and also have edge-cases to deal with in future. (should a file be name `plank__for.overlay` (equivalent to `BOARD=plank//foo`) or just `plank_foo.overlay` ?

The possibility to omit SoC on command line invocation is a user convenience, but there is no reason why a fixed file name should be allowed to do so, because naming a file is a one time operation. 

We seem to have accepted that the ability to inherit overlay files was a feature with undesired side-effects, and therefore long term there is no reason to allow different ways to name the same file.

Of course we don't need to print the warning immediately, but can allow a longer grace period for transitioning the overlay file names.

Reply:
So before I start working on this I want agreement from other that this is going to be accepted, the following are all board identifiers which do not have the soc in the name, so all these board identifiers will need to change, and any overlay or Kconfig fragments for these will need to be renamed:
```
96b_aerocore2
96b_argonkey
96b_avenger96
96b_neonkey
96b_nitrogen
96b_stm32_sensor_mez
96b_wistrio
acn52832
acrn
acrn_ehl_crb
adafruit_feather_m0_basic_proto
adafruit_feather_m0_lora
adafruit_feather_stm32f405
adafruit_grand_central_m4_express
adafruit_itsybitsy_m4_express
adafruit_kb2040
nrf52_adafruit_feather
adafruit_qt_py_rp2040
adafruit_trinket_m0
adi_eval_adin1110ebz
adi_eval_adin2111ebz
adi_sdp_k1
pandora_stm32l475
altera_max10
apollo4p_blue_kxr_evb
apollo4p_evb
kv260_r5
arduino_due
arduino_mkrzero
arduino_nano_33_ble
arduino_nano_33_iot
arduino_nicla_sense_me
arduino_uno_r4_minima
arduino_zero
fvp_base_revc_2xaemv8a
v2m_beetle
ast1030_evb
sam4e_xpro
sam4l_ek
sam4s_xplained
samc21n_xpro
samd20_xpro
samd21_xpro
same54_xpro
saml21_xpro
samr21_xpro
samr34_xpro
bbc_microbit
bbc_microbit_v2
beagleconnect_freedom
beaglev_fire
swan_r5
bcm958401m2
xt-sim
cy8ckit_062s4
cy8cproto_062_4343w
cy8cproto_063_ble
zybo
dragino_lsn50
dragino_nbsn95
titanium_ti60_f225
mercury_xu
kb1200_evb
esp32c3_devkitm
esp32s2_saola
fk7b0m1_vbt6
roc_rk3568_pc
esp32s2_franzininho
generic_leon3
gr716a_mini
gd32a503v_eval
gd32e103v_eval
gd32e507v_start
gd32e507z_eval
gd32f350r_eval
gd32f403z_eval
gd32f407v_start
gd32f450i_eval
gd32f450v_start
gd32f450z_eval
gd32f470i_eval
gd32l233r_eval
gd32vf103c_starter
gd32vf103v_eval
google_dragonclaw
google_kukui
google_twinkie_v2
holyiot_yj16019
xmc45_relax_kit
xmc47_relax_kit
intel_adl_crb
intel_adl_rvp
intel_ehl_crb
intel_ish_5_4_1
intel_ish_5_6_0
intel_ish_5_8_0
niosv_g
niosv_m
intel_rpl_p_crb
intel_rpl_s_crb
intel_socfpga_agilex5_socdk
intel_socfpga_agilex_socdk
cyclonev_socdk
it82xx2_evb
it8xxx2_evb
khadas_edgev
bl652_dvk
bl653_dvk
bl654_dvk
bl654_sensor_board
bl654_usb
bt510
bt610
mg100
pinnacle_100_dvk
rm1xx_dvk
opentitan_earlgrey
stamp_c3
mm_feather
mm_swiftio
ev11l78a
m2gl025_miv
mec1501modular_assy6885
mec15xxevb_assy6853
mec172xevb_assy6906
mec172xmodular_assy6930
mpfs_icicle
mikroe_clicker_2
mikroe_mini_m4_for_stm32
az3166_iotdevkit
native_posix
native_sim
nrf52_bsim
nrf5340bsim_nrf5340_cpuapp
nrf5340bsim_nrf5340_cpunet
npcx4m8f_evb
npcx7m6fb_evb
npcx9m6f_evb
numaker_pfm_m467
numaker_pfm_m487
numaker_m2l31ki
frdm_k22f
frdm_k64f
frdm_k82f
frdm_kl25z
frdm_kw41z
lpcxpresso11u68
lpcxpresso51u68
lpcxpresso55s06
lpcxpresso55s16
lpcxpresso55s28
lpcxpresso55s36
ls1046ardb
mimxrt1010_evk
mimxrt1015_evk
mimxrt1020_evk
mimxrt1024_evk
mimxrt1040_evk
mimxrt1050_evk
mimxrt1050_evk@qspi
mimxrt1060_evk@hyperflash
mimxrt1060_evk
mimxrt1060_evkb
mimxrt1062_fmurt6
mimxrt1064_evk
mimxrt685_evk
mr_canhubk3
rd_rw612_bga
rddrone_fmuk66
twr_ke18f
twr_kv58f220m
ucans32k1sic
usb_kw24d512
olimex_lora_stm32wl_devkit
olimex_lora_stm32wl_devkit@D
olimexino_stm32
olimex_stm32_e407
olimex_stm32_h103
olimex_stm32_h405
olimex_stm32_h407
olimex_stm32_p405
black_f407ve
black_f407zg_pro
icev_wireless
neorv32
stm32_min_dev@black
stm32_min_dev@blue
stm32f030_demo
stm32f401_mini
serpente
pan1770_evb
pan1780_evb
pan1781_evb
pan1782_evb
particle_argon
particle_boron
nrf51_blenano
nrf52_blenano2
particle_xenon
reel_board
reel_board@2
pinetime_devkit0
teensy40
teensy41
qemu_cortex_a53
qemu_cortex_a9
qemu_cortex_m0
qemu_cortex_m3
qemu_cortex_r5
qemu_kvm_arm64
qemu_leon3
qemu_malta
qemu_nios2
qemu_riscv32
qemu_riscv32_xip
qemu_riscv32e
qemu_riscv64
qemu_x86
qemu_x86_64
qemu_x86_lakemont
qemu_x86_tiny
qemu_xtensa
decawave_dwm1001_dev
qomu
quick_feather
rpi_4b
rpi_pico
da14695_dk_usb
da1469x_dk_pro
rcar_salvator_xs
rcar_spider_s4
rzt2m_starter_kit
riscv32_virtual
ronoth_lodev
ruuvi_ruuvitag
scobc_module1
faze
legend@25hdd
legend@25ssd
legend@35
stm32f3_seco_d23
lora_e5_dev_board
lora_e5_mini
seeeduino_xiao
wio_terminal
xiao_ble
xiao_esp32c3
ip_k66f
segger_trb_stm32f407
hifive1
hifive1@B
hifive_unleashed
hifive_unmatched
efm32gg_sltb009a
efm32gg_slwstk6121a
efm32gg_stk3701a
efm32hg_slstk3400a
efm32pg_stk3401a
efm32wg_stk3800
efr32bg22_brd4184a
efr32bg22_brd4184b
efr32bg27_brd2602a
efr32mg_sltb004a
efr32xg24_dk2601b
longan_nano
nrf52_sparkfun
sparkfun_pro_micro_rp2040
sparkfun_red_v_things_plus
b_g474e_dpow1
b_l072z_lrwan1
b_l4s5i_iot01a
b_u585i_iot02a
disco_l475_iot1
nucleo_c031c6
nucleo_f030r8
nucleo_f030r8@2
nucleo_f031k6
nucleo_f042k6
nucleo_f070rb
nucleo_f091rc
nucleo_f103rb
nucleo_f207zg
nucleo_f302r8
nucleo_f303k8
nucleo_f303re
nucleo_f334r8
nucleo_f401re
nucleo_f410rb
nucleo_f411re
nucleo_f412zg
nucleo_f413zh
nucleo_f429zi
nucleo_f446re
nucleo_f446ze
nucleo_f722ze
nucleo_f746zg
nucleo_f756zg
nucleo_f767zi
nucleo_g031k8
nucleo_g070rb
nucleo_g071rb
nucleo_g0b1re
nucleo_g431rb
nucleo_g474re
nucleo_h563zi
nucleo_h723zg
nucleo_h743zi
nucleo_h753zi
nucleo_h7a3zi_q
nucleo_l011k4
nucleo_l031k6
nucleo_l053r8
nucleo_l073rz
nucleo_l152re
nucleo_l412rb_p
nucleo_l432kc
nucleo_l433rc_p
nucleo_l452re
nucleo_l476rg
nucleo_l496zg
nucleo_l4a6zg
nucleo_l4r5zi
nucleo_l552ze_q
nucleo_u575zi_q
nucleo_u5a5zj_q
nucleo_wb55rg
nucleo_wba52cg
nucleo_wba55cg
nucleo_wl55jc
sensortile_box
sensortile_box_pro
st25dv_mb1283_disco
steval_fcu001v1
stm3210c_eval
stm32373c_eval
stm32f072_eval
stm32f072b_disco
stm32f0_disco
stm32f103_mini
stm32f3_disco
stm32f3_disco@E
stm32f411e_disco@B
stm32f411e_disco
stm32f412g_disco
stm32f429i_disc1
stm32f469i_disco
stm32f4_disco
stm32f723e_disco
stm32f746g_disco
stm32f7508_dk
stm32f769i_disco
stm32g0316_disco
stm32g071b_disco
stm32g081b_eval
stm32h573i_dk
stm32h735g_disco
stm32h750b_dk
stm32h7b3i_dk
stm32l1_disco
stm32l476g_disco
stm32l496g_disco
stm32l4r9i_disco
stm32l562e_dk
stm32mp157c_dk2
stm32u5a9j_dk
stm32vl_disco
stm32wb5mm_dk
stm32wb5mmg
visionfive2
robokit1
tlsr9518adk80d
cc1352p1_launchxl
cc1352r1_launchxl
cc1352r_sensortag
cc26x2r1_launchxl
cc3220sf_launchxl
cc3235sf_launchxl
msp_exp432p401r_launchxl
nrf51_ble400
waveshare_open103z
blackpill_f401cc
blackpill_f401ce
blackpill_f411ce
weact_stm32g431_core
esp32s2_lolin_mini
w5500_evb_pico
xenvm
degu_evk
contextualelectronics_abc
nrf52840_blip
nrf52840_papyr
litex_vexriscv
nrf52832_mdk
nrf52840_mdk
nrf52840_mdk_usb_dongle
hsdk
hsdk4xd
iotdk
up_squared
up_squared_pro_7000
nrf51_vbluno51
nrf52_vbluno52
```

Reply:
@nandojve @jhedberg @golowanow @rerickson1 @dleach02 @dbaluta @hakehuang @erwango for comment on this, might impact on-device testing

Reply:
Why do you quote "only" these targets ? is it because they are today used in samples.yaml and tests.yaml?

If yes, does it mean that basically, you impose the use of `<board>/<soc>` syntax ?

Reply:
> Why do you quote "only" these targets ? is it because they are today used in samples.yaml and tests.yaml?
> 
> If yes, does it mean that basically, you impose the use of `<board>/<soc>` syntax ?

It would mean having them as the identifier for usage in twister (forced here) and in filenames for overlays e.g. using the example of `nucleo_l031k6` the identifier for twister (and in any testcase.yaml or sample.yaml files) would need to have `nucleo_l031k6/stm32l031xx`, and you might have an application with a `boards/nucleo_l031k6.conf` file, that file name would be deprecated and you would need to use `boards/nucleo_l031k6_stm32l031xx.conf`. The one thing that would not change would be how you use it to build on the command line, so the soc can be omitted as it on supports one soc, thus building using `cmake -GNinja -DBOARD=nucleo_l031k6 <app>` will work

Reply:
This is not part of the PR (yet) just this discussion thread, want to ensure people are OK with it being added before working on it

Reply:
Maybe a usual RFC would be more appropriate to get feedback.
This should sound obvious to you and you probably want to move fast but we're all working on other stuff and we're not up to speed with daily hwmv2 evolutions.

Reply:
So, on my side, I'm finding it confusing that we could have targets and twister references that can differ. I expect lot of confusion here, so I'm not in favor of it. What are the options?

Reply:
5th commit here shows the changes

Reply:
I agreed with @erwango and I'm favor to get the things without any confusing. This means that I would prefer go verbose and have a clear picture about how the overlays etc are selected and how is the order of "merging" content. I don't want to try guess how that could possible work, it should be precise and clear since beginning.

Maybe it could be a little overkill for single core cpu but the complexity grows to much for Multi-core Asymmetric SoC's with TFM for instance. BTW, this new model may allow the development of SOM (Systems On Module) and than we can have multi CPUs etc.

So, if it is necessary to drop complete the hwmv1 compatibility I'm in favor of that too. I think this is a big change and since we are moving to a LTS I prefer not think about hwmv1 anymore, if that is necessary to have a very clear procedure.